endpoint.go

```diff
@@ -31,14 +30,12 @@ type Endpoint interface {
 	// Network returns the name of the network to which this endpoint is attached.
 	Network() string
 
-	// Join creates a new sandbox for the given container ID and populates the
-	// network resources allocated for the endpoint and joins the sandbox to
-	// the endpoint.
-	Join(containerID string, options ...EndpointOption) error
+	// Join joins the sandbox to the endpoint and populates into the sandbox
+	// the network resources allocated for the endpoint.
+	Join(sandbox Sandbox, options ...EndpointOption) error
```

Comment:
As a libnetwork client, I get objects of three kinds (Network, Endpoint, Sandbox) and I am expecting to use APIs which takes these three blocks as parameters. I see these as my three kind of lego block and use them.
I do not feel giving API which take a property (in this case ID) of the entity makes it clean.

Also, this new way it is clear to the user that if you want to use a Join API, you first need the Sandbox ingredient. From the API signature, the flow is now controlled and straightforward.
Instead if Join takes an ID, it is not straightforward from where that id has to come from. 

When we were not exposing the Sandbox entity, id was the only solution. Now it is no longer true. Only who holds the instance of Sandbox can call this API.